### PR TITLE
Don't check [print/write]_with_newline on raw strings

### PR DESCRIPTION
**Pull request by uniphil**
Some tests for #3778 and some maybe-not-the-greatest code that passes those tests!

I didn't run `fmt` because a) it doesn't seem to install on nightly for me, and b) on stable it wanted to apply formatting to over 90 files. Happy to make any tweaks though!

I suspect this contribution may require more than just tweaks. I'm still sort of new to rust so it may not be idiomatic, and the specific approach I took feels a little heavy-handed and brittle. I'm happy to make changes with some guidance, or equally happy if this gives a starting place for someone else to do it better :)

### CODE REVIEW

File: clippy_lints/src/write.rs [hunk 1]
```diff
@@ -212,8 +212,8 @@ impl EarlyLintPass for Pass {
                 }
             }
         } else if mac.node.path == "write" {
-            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, true).0 {
-                if check_newlines(&fmtstr) {
+            if let (Some(fmtstr), _, is_raw) = check_tts(cx, &mac.node.tts, true) {
+                if !is_raw && check_newlines(&fmtstr) {
```
Comment (oli-obk) [inline clippy_lints/src/write.rs]: I think we could pass the `is_raw` into `check_newlines` and give that function the ability to detect

```rust
write!(foo, "
");
```

and

```rust
write!(foo, r"
");
```

Comment (uniphil) [reply clippy_lints/src/write.rs]: sure! implemented in the latest push.

**Comment by uniphil** [comment]
I added tests and code to detect literal terminating newlines

```rust
write!(foo, "
");
write!(foo, r"
");
print!("
");
print!(r"
");
```

I couldn't find it but I was wondering -- is there a way to take a string with literal escape sequences (our `fmtstr` from `check_tts`) and apply/"unescape" them? That whole function could basically become

```rust
if s.ends_with('\n') {
    return true;
}
```

**Comment by oli-obk** [comment]
> I couldn't find it but I was wondering -- is there a way to take a string with literal escape sequences (our `fmtstr` from `check_tts`) and apply/"unescape" them? That whole function could basically become

I checked, but couldn't find anything reusable. I think it's used inline in https://github.com/rust-lang/rust/blob/862ebc4c38958b8ba5e30dd5eb1203e6109c40a2/src/libsyntax_ext/format.rs#L808

you still have some travis failures

**Comment by uniphil** [comment]
> > I couldn't find it but I was wondering -- is there a way to take a string with literal escape sequences (our `fmtstr` from `check_tts`) and apply/"unescape" them? That whole function could basically become
> 
> I checked, but couldn't find anything reusable. I think it's used inline in https://github.com/rust-lang/rust/blob/862ebc4c38958b8ba5e30dd5eb1203e6109c40a2/src/libsyntax_ext/format.rs#L808

ah ok. too bad!

> you still have some travis failures

reformatting. I'll get it when I get home tonight.

**Comment by oli-obk** [comment]
@bors r+

Thanks!

**Pull request merged**